src/operator/nn/mkldnn/mkldnn_rnn_impl.h

```diff
@@ -0,0 +1,589 @@
+  stream(stream::kind::eager).submit({r}).wait();
+}
+
+template <typename DType>
+mkldnn::memory ConcatData(mkldnn::memory::format src_format,
+                          mkldnn::memory::format dst_format,
+                          std::vector<mkldnn::memory::dims> srcs_cds,
+                          mkldnn::memory::dims dst_cds,
+                          mkldnn::memory::data_type mkldnn_dtype,
+                          int concat_dimension,
+                          std::vector<DType*> srcs_data) {
+  auto cpu_engine = CpuEngine::Get()->get_engine();
+  std::vector<mkldnn::memory::primitive_desc> srcs_pd;
+  std::vector<mkldnn::memory> srcs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    auto desc = memory::desc(srcs_cds[i], mkldnn_dtype, src_format);
+    auto mpd = memory::primitive_desc(desc, cpu_engine);
+    auto src_memory = mkldnn::memory({desc, cpu_engine}, srcs_data[i]);
+    srcs_pd.push_back(mpd);
+    srcs.push_back(src_memory);
+  }
+  std::vector<primitive::at> inputs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    inputs.push_back(srcs[i]);
+  }
+
+  auto dst_desc = mkldnn::memory::desc(dst_cds, mkldnn_dtype, dst_format);
+  auto concat_pd = mkldnn::concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  auto dst = mkldnn::memory(concat_pd.dst_primitive_desc());
+  
+  auto c = mkldnn::concat(concat_pd, inputs, dst);
+  stream(stream::kind::eager).submit({c}).wait();
+  return dst;
+}
+
+template <typename DType>
+void MKLDNNRNNForward(bool state_outputs,
+                      int L,
+                      int D,
+                      const int T,
+                      const int N,
+                      int I,
+                      const int H,
+                      DType* x_ptr,
+                      DType* hx_ptr,
+                      DType* cx_ptr,
+                      DType* w_ptr,
+                      DType* b_ptr,
+                      DType* y_ptr,
+                      DType* hy_ptr,
+                      DType* cy_ptr,
+                      std::vector<mkldnn::memory> *weight_bias_memory,
+                      int mode) {
+  int ngates = 0, nstates = 0;
+  mkldnn::algorithm nalgorithm = GetMKLDNNAlgo(mode, &ngates, &nstates);
+  mkldnn::memory::data_type mkldnn_dtype = GetMKLDNNDataType<DType>();
+  const int b_size = 2 * H * ngates * D;
+  const int cell_size = N * H * D;
+  const int single_cell_size = N * H;
+  const int single_b_size = ngates * H;
+  //  First layer
+  int w_size = (I + H) * H * ngates * D;
+  DType* x_0 = x_ptr;  //  T * N * I
+  DType* y = y_ptr;  //  T * N * D * H
+  DType* back_w_ptr = w_ptr;
+  DType* back_b_ptr = b_ptr;
+  DType* wx_0 = w_ptr;  //  ngates * H, I
+  DType* wh_0 = w_ptr + I * H * ngates;  //  ngates * H, H
+  if (D == 2) {
+    back_w_ptr = w_ptr + ngates * H * (I + H);
+    back_b_ptr = b_ptr + single_b_size * 2;
+  }
+  DType* back_wx_0 = back_w_ptr;
+  DType* back_wh_0 = back_w_ptr + I * H * ngates;
```

Comment:
move all the assignment of back_xxx to `if (D == 2)` closure, and with default value here is enough.